src/Neo/SmartContract/ApplicationEngine.Runtime.cs

```diff
@@ -395,8 +400,15 @@ protected internal void RuntimeNotifyV1(byte[] eventName, Array state)
         protected internal void SendNotification(UInt160 hash, string eventName, Array state)
         {
             NotifyEventArgs notification = new(ScriptContainer, hash, eventName, (Array)state.DeepCopy(asImmutable: true));
-            Notify?.Invoke(this, notification);
             notifications ??= new List<NotifyEventArgs>();
+            // Restrict the number of notifications for Application executions. Do not check
+            // persisting triggers to avoid native persist failure. Do not check verification
+            // trigger since verification context is loaded with ReadOnly flag.
+            if (IsHardforkEnabled(Hardfork.HF_Echidna) && Trigger == TriggerType.Application && notifications.Count == MaxNotificationCount)
```

Comment:
Looks good to me, we just need to ensure .Count is correct calculated at this step.

Reply:
Is trigger really needed? If it's an limit shouldn't have a trigger.

Reply:
Yes, it's needed, because we can't allow OnPersist and PostPersist failures (imagine that in future MaxNotificationCount may be reduced to some lower value). And for Verification trigger this check is useless by design, no notifications are possible with this trigger.

Reply:
Ref. https://github.com/nspcc-dev/neo-go/issues/3490#issuecomment-2437272527.